File: mobile/android/app/src/main/java/com/volttracker/obdpoc/ElmConnection.java
package com.volttracker.obdpoc;

import android.bluetooth.BluetoothDevice;
import android.bluetooth.BluetoothSocket;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.nio.charset.StandardCharsets;
import java.util.UUID;
import java.util.function.BooleanSupplier;

/**
 * Owns the Bluetooth RFCOMM socket to the ELM327 adapter and the low-level command IO. Extracted
 * from {@link ObdService} so socket/stream handling is isolated and the {@link #transact} read loop
 * is unit-testable against in-memory streams.
 *
 * <p>Not {@code final}: {@code ObdPollingEngineTest} subclasses this to script adapter responses
 * (overriding {@link #open}, {@link #transact}, {@link #sendEscape}, {@link #close}) so the
 * polling-engine state machine can be exercised without a real RFCOMM socket.
 */
class ElmConnection {

    /**
     * Indirected wall clock — tests inject a synthetic clock so {@link #transact}'s deadline check
     * is deterministic rather than asserting on wall time (CI-flaky). Defined as a package-private
     * interface rather than {@code java.util.function.LongSupplier} so we don't need API 24 / core
     * library desugaring (minSdk is 23).
     */
    interface Clock {
        long nowMs();
    }

    /**
     * Per-phase wall-clock timings (ms) for the last successful {@link #open} call. Field-readable
     * so {@link ObdPollingEngine} can stamp these onto the {@code socket_open_result} event without
     * needing a side-channel — A1/B3 in the connection-hardening plan.
     *
     * <p>{@link #firstReadMs} is populated by {@link #wakeNudge} (the first read after open) rather
     * than by {@link #open} itself, because the OS rarely surfaces a wedged adapter until the first
     * byte is requested; setting it during {@code wakeNudge} keeps the {@code first_read} label on
     * the right operation. All three default to {@code -1} when not measured.
     */
    long rfcommConnectMs = -1L;

    long getStreamsMs = -1L;
    long firstReadMs = -1L;

    /**
     * Phase the most recent {@link #open}/{@link #wakeNudge} failure occurred in. Mirrors the
     * {@code errorPhase} field on the {@code socket_open_result} event. One of {@code
     * "rfcomm_connect"}, {@code "get_streams"}, {@code "first_read"}, or empty when the call
     * succeeded.
     */
    String lastErrorPhase = "";

    private BluetoothSocket socket;
    private InputStream input;
    private OutputStream output;
    private final Clock clock;

    ElmConnection() {
        this.clock = System::currentTimeMillis;
    }

    /** Test constructor: drives {@link #transact}/{@link #sendEscape} against in-memory streams. */
    ElmConnection(InputStream input, OutputStream output) {
        this(input, output, System::currentTimeMillis);
    }

    /** Test constructor: in-memory streams + injected clock for deterministic timeout tests. */
    ElmConnection(InputStream input, OutputStream output, Clock clock) {
        this.input = input;
        this.output = output;
        this.clock = clock;
    }

    boolean isOpen() {
        return input != null && output != null;
    }

    /**
     * Opens an RFCOMM socket to {@code device}. {@link BluetoothSocket#connect()} has no timeout
     * and can block indefinitely on a dead adapter, so a daemon watchdog closes the socket after
     * {@code connectTimeoutMs}, which makes the blocked connect throw IOException and lets the
     * normal failure path take over.
     */
    void open(BluetoothDevice device, UUID uuid, long connectTimeoutMs) throws IOException {
        rfcommConnectMs = -1L;
        getStreamsMs = -1L;
        firstReadMs = -1L;
        lastErrorPhase = "";

        BluetoothSocket pendingSocket = device.createRfcommSocketToServiceRecord(uuid);
        socket = pendingSocket;
        Thread watchdog =
                new Thread(
                        () -> {
                            sleep(connectTimeoutMs);
                            if (!pendingSocket.isConnected()) {
                                try {
                                    pendingSocket.close();
                                } catch (IOException ignored) {
                                }
                            }
                        });
        watchdog.setDaemon(true);
        watchdog.start();
        long connectStart = clock.nowMs();
        try {
            pendingSocket.connect();
            rfcommConnectMs = Math.max(0L, clock.nowMs() - connectStart);
        } catch (IOException ex) {
            rfcommConnectMs = Math.max(0L, clock.nowMs() - connectStart);
            lastErrorPhase = "rfcomm_connect";
            throw ex;
        } finally {
            watchdog.interrupt();
        }
        long streamsStart = clock.nowMs();
        try {
            input = socket.getInputStream();
            output = socket.getOutputStream();
            getStreamsMs = Math.max(0L, clock.nowMs() - streamsStart);
        } catch (IOException ex) {
            getStreamsMs = Math.max(0L, clock.nowMs() - streamsStart);
            lastErrorPhase = "get_streams";
            throw ex;
        }
    }

    /** Result of a {@link #wakeNudge} call — recorded onto the {@code wake_nudge} JSONL event. */
    static final class WakeNudgeResult {
        final long durationMs;
        final boolean gotResponse;

        WakeNudgeResult(long durationMs, boolean gotResponse) {
            this.durationMs = durationMs;
            this.gotResponse = gotResponse;
        }
    }

    /**
     * Sends a single {@code \r} no-op and waits up to {@code toleranceMs} for any response byte.
     * Some adapters (notably wedged OBDLink MX+ units) need a no-op to spin up before {@code ATZ}
     * lands; on a healthy adapter this returns immediately with {@code gotResponse=true} and on a
     * wedged adapter it returns with {@code gotResponse=false}, which the engine logs as a {@code
     * wake_nudge} event for later triage.
     *
     * <p>Doubles as the {@code first_read} timing source: a failing {@link InputStream#available()}
     * /{@link InputStream#read} surfaces the wedged-adapter symptom that motivated the bucket
     * split. On {@link IOException} the call sets {@link #firstReadMs} + {@link #lastErrorPhase}
     * and re-throws so the engine's failure path can classify it.
     */
    WakeNudgeResult wakeNudge(long toleranceMs) throws IOException {
        if (output == null || input == null) {
            throw new IOException("Adapter stream is not open");
        }
        long start = clock.nowMs();
        boolean gotResponse = false;
        try {
            output.write((byte) '\r');
            output.flush();
            long deadline = start + Math.max(0L, toleranceMs);
            byte[] buffer = new byte[64];
            while (clock.nowMs() < deadline) {
                int available = input.available();
                if (available > 0) {
                    int read = input.read(buffer, 0, Math.min(buffer.length, available));
                    if (read > 0) {
                        gotResponse = true;
                        break;
                    }
                } else {
                    sleep(20);
                }
            }
            firstReadMs = Math.max(0L, clock.nowMs() - start);
            return new WakeNudgeResult(firstReadMs, gotResponse);
        } catch (IOException ex) {
            firstReadMs = Math.max(0L, clock.nowMs() - start);
            lastErrorPhase = "first_read";
            throw ex;
        }
    }

    /**
     * Writes {@code command} and reads the reply until the ELM '>' prompt, the timeout, or {@code
     * keepWaiting} going false. Returns the raw response (possibly partial).
     *
     * @throws IOException if the stream is not open or the socket has broken
     */
    String transact(String command, long timeoutMs, BooleanSupplier keepWaiting)
            throws IOException {
        if (output == null || input == null) {
            throw new IOException("Adapter stream is not open");
        }
        drainInput();
        output.write((command + "\r").getBytes(StandardCharsets.US_ASCII));
        output.flush();

        StringBuilder response = new StringBuilder();
        long deadline = clock.nowMs() + timeoutMs;
        byte[] buffer = new byte[128];
        while (clock.nowMs() < deadline && keepWaiting.getAsBoolean()) {
            int available = input.available();
            if (available > 0) {
                int read = input.read(buffer, 0, Math.min(buffer.length, available));
                if (read > 0) {
                    String chunk = new String(buffer, 0, read, StandardCharsets.US_ASCII);
                    response.append(chunk);
                    if (chunk.indexOf('>') >= 0) {
                        break;
                    }
                }
            } else {
                sleep(25);
            }
        }
        return response.toString();
    }

    /** Sends the ELM escape byte and drains whatever the adapter echoes back. */
    void sendEscape(long settleMs) throws IOException {
        if (output == null || input == null) {
            return;
        }
        output.write(0x1B);
        output.flush();
        sleep(settleMs);
        drainInput();
    }

    void close() {
        try {
            if (input != null) {
                input.close();
            }
        } catch (IOException ignored) {
        }
        try {
            if (output != null) {
                output.close();
            }
        } catch (IOException ignored) {
        }
        try {
            if (socket != null) {
                socket.close();
            }
        } catch (IOException ignored) {
        }
        input = null;
        output = null;
        socket = null;
    }

    private void drainInput() throws IOException {
        if (input == null) {
            return;
        }
        byte[] buffer = new byte[128];
        int drained = 0;
        // Cap the drain so a chatty/garbage adapter cannot spin this loop forever.
        while (input.available() > 0 && drained < 8192) {
            int read = input.read(buffer, 0, Math.min(buffer.length, input.available()));
            if (read < 0) {
                break;
            }
            drained += read;
        }
    }

    private static void sleep(long millis) {
        try {
            Thread.sleep(millis);
        } catch (InterruptedException ex) {
            Thread.currentThread().interrupt();
        }
    }
}

File: mobile/android/app/src/main/kotlin/com/volttracker/obdpoc/ElmConnection.kt
package com.volttracker.obdpoc

import android.bluetooth.BluetoothDevice
import android.bluetooth.BluetoothSocket
import androidx.annotation.VisibleForTesting
import java.io.IOException
import java.io.InputStream
import java.io.OutputStream
import java.nio.charset.StandardCharsets
import java.util.UUID

/**
 * Owns the Bluetooth RFCOMM socket to the ELM327 adapter and the low-level command IO. Extracted
 * from [ObdService] so socket/stream handling is isolated and [transact] is unit-testable against
 * in-memory streams.
 *
 * Open for `ObdPollingEngineTest`, which subclasses this to script adapter responses without a real
 * RFCOMM socket.
 */
open class ElmConnection
    @JvmOverloads
    constructor(
        private var input: InputStream? = null,
        private var output: OutputStream? = null,
        private val clock: Clock = Clock { System.currentTimeMillis() },
        private val socketFactory: SocketFactory =
            SocketFactory { device, uuid ->
                BluetoothElmSocket(device.createRfcommSocketToServiceRecord(uuid))
            },
    ) {
        fun interface Clock {
            fun nowMs(): Long
        }

        fun interface SocketFactory {
            fun create(
                device: BluetoothDevice,
                uuid: UUID,
            ): ElmSocket
        }

        fun interface KeepWaiting {
            fun getAsBoolean(): Boolean
        }

        interface ElmSocket {
            val isConnected: Boolean
            val inputStream: InputStream
            val outputStream: OutputStream

            @Throws(IOException::class)
            fun connect()

            @Throws(IOException::class)
            fun close()
        }

        @JvmField var rfcommConnectMs: Long = -1L

        @JvmField var getStreamsMs: Long = -1L

        @JvmField var firstReadMs: Long = -1L

        @JvmField var lastErrorPhase: String = ""

        @Volatile @JvmField
        var watchdogFired: Boolean = false

        @JvmField var lastTransactTruncated: Boolean = false

        private var socket: ElmSocket? = null

        fun isOpen(): Boolean = input != null && output != null

        /**
         * Opens an RFCOMM socket to [device]. `BluetoothSocket.connect()` has no timeout and can
         * block indefinitely on a dead adapter, so a daemon watchdog closes the socket after
         * [connectTimeoutMs].
         *
         * The backticked name reads naturally at call sites (connection.open(...)) despite
         * colliding with Kotlin's soft keyword; not worth renaming a stable seam over.
         */
        @Suppress("FunctionNaming")
        @Throws(IOException::class)
        open fun `open`(
            device: BluetoothDevice,
            uuid: UUID,
            connectTimeoutMs: Long,
        ) {
            openSocket(socketFactory.create(device, uuid), connectTimeoutMs)
        }

        @Throws(IOException::class)
        @VisibleForTesting
        internal fun openSocketForTest(
            pendingSocket: ElmSocket,
            connectTimeoutMs: Long,
        ) {
            openSocket(pendingSocket, connectTimeoutMs)
        }

        private fun openSocket(
            pendingSocket: ElmSocket,
            connectTimeoutMs: Long,
        ) {
            rfcommConnectMs = -1L
            getStreamsMs = -1L
            firstReadMs = -1L
            lastErrorPhase = ""
            watchdogFired = false
            // Per-session state: a fresh socket must not inherit the previous session's
            // truncation verdict (readers consult it before the first transact completes).
            lastTransactTruncated = false

            socket = pendingSocket
            val watchdog =
                Thread {
                    try {
                        Thread.sleep(connectTimeoutMs)
                    } catch (ex: InterruptedException) {
                        Thread.currentThread().interrupt()
                        return@Thread
                    }
                    if (!pendingSocket.isConnected) {
                        watchdogFired = true
                        try {
                            pendingSocket.close()
                        } catch (ex: IOException) {
                            OBDLog.warn("ElmConnection", "watchdog socket close failed: ${ex.message}")
                        }
                    }
                }
            watchdog.isDaemon = true
            watchdog.start()

            val connectStart = clock.nowMs()
            try {
                pendingSocket.connect()
                rfcommConnectMs = maxOf(0L, clock.nowMs() - connectStart)
            } catch (ex: IOException) {
                rfcommConnectMs = maxOf(0L, clock.nowMs() - connectStart)
                lastErrorPhase = "rfcomm_connect"
                throw ex
            } finally {
                watchdog.interrupt()
            }

            val streamsStart = clock.nowMs()
            try {
                input = socket?.inputStream
                output = socket?.outputStream
                getStreamsMs = maxOf(0L, clock.nowMs() - streamsStart)
            } catch (ex: IOException) {
                getStreamsMs = maxOf(0L, clock.nowMs() - streamsStart)
                lastErrorPhase = "get_streams"
                throw ex
            }
        }

        /** Result of a [wakeNudge] call, recorded onto the `wake_nudge` JSONL event. */
        class WakeNudgeResult(
            @JvmField val durationMs: Long,
            @JvmField val gotResponse: Boolean,
        )

        /**
         * Sends a single no-op carriage return and waits up to [toleranceMs] for any response byte.
         * Doubles as the first-read timing source.
         */
        @Throws(IOException::class)
        open fun wakeNudge(toleranceMs: Long): WakeNudgeResult {
            val out = output ?: throw IOException("Adapter stream is not open")
            val inputStream = input ?: throw IOException("Adapter stream is not open")
            val start = clock.nowMs()
            var gotResponse = false
            try {
                out.write('\r'.code)
                out.flush()
                val deadline = start + maxOf(0L, toleranceMs)
                val buffer = ByteArray(64)
                while (clock.nowMs() < deadline) {
                    val available = inputStream.available()
                    if (available > 0) {
                        val read = inputStream.read(buffer, 0, minOf(buffer.size, available))
                        if (read > 0) {
                            gotResponse = true
                            break
                        }
                    } else {
                        sleep(20)
                    }
                }
                firstReadMs = maxOf(0L, clock.nowMs() - start)
                return WakeNudgeResult(firstReadMs, gotResponse)
            } catch (ex: IOException) {
                firstReadMs = maxOf(0L, clock.nowMs() - start)
                lastErrorPhase = "first_read"
                throw ex
            }
        }

        /**
         * Writes [command] and reads the reply until the ELM `>` prompt, the timeout, or
         * [keepWaiting] going false.
         */
        @Throws(IOException::class)
        open fun transact(
            command: String,
            timeoutMs: Long,
            keepWaiting: KeepWaiting,
        ): String {
            val out = output ?: throw IOException("Adapter stream is not open")
            val inputStream = input ?: throw IOException("Adapter stream is not open")
            lastTransactTruncated = false
            drainInput()
            out.write((command + "\r").toByteArray(StandardCharsets.US_ASCII))
            out.flush()

            val response = StringBuilder()
            val startMs = clock.nowMs()
            val deadline = startMs + timeoutMs
            val buffer = ByteArray(128)
            var lastByteAtMs = startMs
            while (clock.nowMs() < deadline && keepWaiting.getAsBoolean()) {
                val available = inputStream.available()
                if (available > 0) {
                    val read = inputStream.read(buffer, 0, minOf(buffer.size, available))
                    if (read > 0) {
                        lastByteAtMs = clock.nowMs()
                        val chunk = String(buffer, 0, read, StandardCharsets.US_ASCII)
                        response.append(chunk)
                        if (chunk.indexOf('>') >= 0) {
                            break
                        }
                    }
                } else {
                    // The '>' prompt hasn't arrived. If a response already came in and the adapter has
                    // since gone quiet for longer than any normal inter-frame gap, the ELM327 v1.4b
                    // almost certainly dropped the prompt — stop here instead of burning the rest of
                    // the timeout, so the caller's prompt-recovery runs now (saves ~1 s+ per drop).
                    if (response.isNotEmpty() && clock.nowMs() - lastByteAtMs >= NO_PROMPT_QUIET_PERIOD_MS) {
                        break
                    }
                    sleep(25)
                }
            }
            val text = response.toString()
            lastTransactTruncated =
                text.isNotEmpty() &&
                text.indexOf('>') < 0 &&
                clock.nowMs() >= deadline
            return text
        }

        /** Sends the ELM escape byte and drains whatever the adapter echoes back. */
        @Throws(IOException::class)
        open fun sendEscape(settleMs: Long) {
            val out = output ?: return
            input ?: return
            out.write(0x1B)
            out.flush()
            sleep(settleMs)
            drainInput()
        }

        open fun close() {
            try {
                input?.close()
            } catch (ex: IOException) {
                OBDLog.warn("ElmConnection", "input stream close failed: ${ex.message}")
            }
            try {
                output?.close()
            } catch (ex: IOException) {
                OBDLog.warn("ElmConnection", "output stream close failed: ${ex.message}")
            }
            try {
                socket?.close()
            } catch (ex: IOException) {
                OBDLog.warn("ElmConnection", "socket close failed: ${ex.message}")
            }
            input = null
            output = null
            socket = null
            lastTransactTruncated = false
        }

        @Throws(IOException::class)
        private fun drainInput() {
            val inputStream = input ?: return
            val buffer = ByteArray(128)
            var drained = 0
            while (inputStream.available() > 0 && drained < 8192) {
                val read = inputStream.read(buffer, 0, minOf(buffer.size, inputStream.available()))
                if (read < 0) {
                    break
                }
                drained += read
            }
        }

        private companion object {
            // Once a response has arrived, this much continued silence with still no '>' prompt means
            // the ELM327 v1.4b dropped the prompt (a known quirk). It is far longer than a normal
            // inter-frame gap (<100 ms), so a legitimate slow multi-frame reply is not truncated; it
            // just lets prompt-recovery start ~1 s+ sooner than waiting out the full command timeout.
            const val NO_PROMPT_QUIET_PERIOD_MS = 250L

            fun sleep(millis: Long) {
                try {
                    Thread.sleep(millis)
                } catch (ex: InterruptedException) {
                    Thread.currentThread().interrupt()
                }
            }
        }

        private class BluetoothElmSocket(
            private val delegate: BluetoothSocket,
        ) : ElmSocket {
            override val isConnected: Boolean
                get() = delegate.isConnected

            override val inputStream: InputStream
                get() = delegate.inputStream

            override val outputStream: OutputStream
                get() = delegate.outputStream

            override fun connect() {
                delegate.connect()
            }

            override fun close() {
                delegate.close()
            }
        }
    }
